audit/audit.go

```diff
@@ -0,0 +1,147 @@
+package audit
+
+import (
+	"crypto"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/tls"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+var emptyHash = [32]byte{}
+
+// B64LeafHash calculates the base64-encoded leaf hash of the certificate or
+// precertificate at chain[0] that sct was issued for.
+//
+// If using this function to calculate the leaf hash for an X509 certificate
+// (i.e. not a precertificate) then it is enough to just provide the end entity
+// certificate in chain.  However, if using this function to calculate the leaf
+// hash for a precertificate then the issuing certificate must also be provided
+// in chain.  When providing a certificate chain the leaf certificate must be at
+// chain[0].
+//
+// sct is required because the SCT timestamp is used to calculate the leaf hash.
+// Leaf hashes are unique to (pre)certificate-SCT pairs.
+func B64LeafHash(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp) (string, error) {
+	hash, err := LeafHash(chain, sct)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(hash[:]), nil
+}
+
+// LeafHash calculates the leaf hash of the certificate or precertificate at
+// chain[0] that sct was issued for.
+//
+// If using this function to calculate the leaf hash for an X509 certificate
+// (i.e. not a precertificate) then it is enough to just provide the end entity
+// certificate in chain.  However, if using this function to calculate the leaf
+// hash for a precertificate then the issuing certificate must also be provided
+// in chain.  When providing a certificate chain the leaf certificate must be at
+// chain[0].
+//
+// sct is required because the SCT timestamp is used to calculate the leaf hash.
+// Leaf hashes are unique to (pre)certificate-SCT pairs.
+func LeafHash(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp) ([32]byte, error) {
+	if len(chain) == 0 {
+		return emptyHash, errors.New("chain is empty")
+	}
+	if sct == nil {
+		return emptyHash, errors.New("sct is nil")
+	}
+
+	certType := ct.X509LogEntryType
+	if IsPrecert(chain[0]) {
+		certType = ct.PrecertLogEntryType
+	}
+	leaf, err := ct.MerkleTreeLeafFromChain(chain, certType, sct.Timestamp)
+	if err != nil {
+		return emptyHash, fmt.Errorf("error creating MerkleTreeLeaf: %s", err)
+	}
+	leafData, err := tls.Marshal(*leaf)
+	if err != nil {
+		return emptyHash, fmt.Errorf("error tls-encoding MerkleTreeLeaf: %s", err)
+	}
+
+	data := append([]byte{0}, leafData...)
+	leafHash := sha256.Sum256(data)
+	return leafHash, nil
+}
+
+// IsPrecert checks whether the given certificate is a precertificate.
+func IsPrecert(cert *x509.Certificate) bool {
+	for _, ext := range cert.Extensions {
+		if ext.Id.Equal(x509.OIDExtensionCTPoison) {
+			return true
+		}
+	}
+	return false
+}
+
+// VerifySCT takes the public key of a Certificate Transparency Log, a
```

Comment:
I'm a bit confused by what's intended here, but I may have misunderstood the general intent.

This *seems* to be aiming at the scenario of checking an SCT just after submission, where the user may have just submitted either a normal certificate or a pre-certificate (and in the latter case, the user is probably a CA).   If that is the case, then the `client` code for submission does already allow immediate SCT checking.

For checking SCTs after the fact, a normal user is always going to be dealing with a certificate rather than an (explicit) precert, and there are two cases:
 - the SCT was for the certificate as-is (implying that the SCT arrived via some parallel mechanism like stapling)
 - the SCT is one that's embedded in the certificate as an SCTList extension.

In the latter case, SCT verification does implicitly need to deal with a pre-cert, but from the perspective of the library user it would be nice to hide that (as dealing with pre-certs is pesky).   So a `VerifyEmbeddedSCT(logKey, chain, sct)` could then take on the work of:
 - building a pre-certificate from the cert by dropping the SCTList extension, being careful not to reorder anything
 - building the Merkle leaf, which requires the cert issuer (in `chain[1]`)
 - checks the signature.

(I guess the latter two parts could be done with your `VerifySCT` here, so maybe that's the aim and I'm just jumping the gun).

Reply:
I've added functionality to be able to handle embedded SCTs too :)